pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -405,6 +420,68 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
 	return out
 }
 
+// mergeIngress will append rules from Ingress-generated virtual services to existing virtual services.
+// The ingress can be created dynamically, and only has very limitted capabilities -
+// [host]/path->backend configuration
+func mergeIngress(configs []model.Config) []*model.Config {
+	configsOut := []*model.Config{}
+
+	hostsToVS := map[string]*networking.VirtualService{}
+
+	// Ignore ingress resources, build a map of hosts to (real) VirtualServices. VirtualService has priority.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if !strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
+				if e {
+					// Just a message - dups will be eliminated later.
+					log.Infof("Duplicated hostname %s OLD:%v NEW:%v", h, vse, vs)
+				}
+				hostsToVS[h] = vs
+				configsOut = append(configsOut, &c)
+			}
+		}
+	}
+
+	// Add the rules from Ingress generated resources, as well as
+	// all ingress generated resources for domains not covered by VirtualService
+
+	// We could restrict ingress merging if a VirtualService exists - only /.well-known (to narrow
+	// it to ACME case, or only exact match, etc.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
+				if e {
+					// A VirtualService entry exists for the domain.
+					for _, r := range vs.Http {
+						if len(r.Match) == 0 {
+							continue // unexpected
+						}
+						if p, ok := r.Match[0].Uri.GetMatchType().(*networking.StringMatch_Prefix); ok {
+							if p.Prefix == "/" || p.Prefix == "" {
+								log.Infof("Ignoring / prefix for %s %s", h, c.Name)
+								continue
+							}
+						}
+						vse.Http = append([]*networking.HTTPRoute{r}, vse.Http...)
```

Comment:
??
This is merging user generated virtual services with user generated ingress. FYI, this is pretty much what we did before 0.8, and there was a never ending stream of bugs. Keeping the entire ingress separate made life much more easier. And we made a very conscious decision to stop with just the conversion and not do any runtime merging with user generated artifacts precisely due to all the bugs we faced. So, lets please not bring this back, one week before 1.0 launch :).

Reply:
Come to think of it, this entire logic is of no use. The gateway code is for a single gateway (istio: ingress, or istio: ingressgateway). So, the list of configs you have here for the current gateway being processed (istio: ingress) are all the auto generated gateways in conversion.go.